File: CPP/oops/bookmyshow/TicketBooking.cpp
#include <iostream>
#include <unordered_map>

class Seat
{
private:
    int seatNo;
    bool reserved;

public:
    Seat(int seatNo) : seatNo(seatNo), reserved(false) {}

    // Method to reserve the seat
    bool reserveSeat()
    {
        if (!reserved)
        {
            reserved = true;
            std::cout << "Seat " << seatNumber << " reserved." << std::endl;
            return true;
        }
        else
        {
            std::cout << "Seat " << seatNumber << " is already reserved." << std::endl;
            return false;
        }
    }

    // Method to unreserve the seat
    bool unreserveSeat()
    {
        if (reserved)
        {
            reserved = false;
            std::cout << "Seat " << seatNumber << " has been successfully unreserved." << std::endl;
            return true;
        }
        else
        {
            std::cout << "Seat " << seatNumber << " is not currently reserved." << std::endl;
            return false;
        }
    }

    bool isReserved() const
    {
        return this->reserved;
    }

    int getSeatNo() const
    {
        return this->seatNo;
    }
};

class SeatManager
{
private:
    std::unordered_map<int, Seat> seats;

public:
    SeatManager(int numberOfSeats)
    {
        for (int i = 1; i <= numberOfSeats; ++i)
        {
            seats.emplace(i, Seat(i)); // Creating seat objects for seat numbers
        }
    }

    void allocateSeat(int seatNumber)
    {
        if (seats.find(seatNumber) != seats.end())
        {
            Seat &seat = seats.at(seatNumber);
            if (!seat.isReserved())
            {
                seat.reserveSeat(); // Allocate the seat if available
            }
            else
            {
                std::cout << "Seat " << seatNumber << " is already reserved. Cannot allocate." << std::endl;
            }
        }
        else
        {
            std::cout << "Invalid seat number: " << seatNumber << std::endl;
        }
    }

    void unreserveSeat(int seatNumber)
    {
        if (seats.find(seatNumber) != seats.end())
        {
            seats.at(seatNumber).unreserveSeat();
        }
        else
        {
            std::cout << "Invalid seat number: " << seatNumber << std::endl;
        }
    }

    // Method to check seat availability
    bool isSeatAvailable(int seatNumber) const
    {
        auto it = seats.find(seatNumber);
        return it != seats.end() && !it->second.isReserved();
    }
};

int main()
{
    // Initialize the SeatManager with 20 seats
    SeatManager seatManager(20);

    // Example: Trying to allocate seats
    seatManager.allocateSeat(5);  // Should reserve seat 5
    seatManager.allocateSeat(5);  // Should show seat 5 is already reserved
    seatManager.allocateSeat(10); // Should reserve seat 10

    // Unreserve a seat
    seatManager.unreserveSeat(5); // Should unreserve seat 5
    seatManager.allocateSeat(5);  // Should reserve seat 5 again after unreserving

    // Trying to reserve an invalid seat
    seatManager.allocateSeat(21); // Should print invalid seat number

    return 0;
}

File: python/practice/OOPs/TicketBooking.py
class Seat:
    def __init__(self, seat_number):
        self.seat_number = seat_number
        self.reserved = False  # Initially, the seat is not reserved.

    # Method to reserve the seat
    def reserve_seat(self):
        if not self.reserved:
            self.reserved = True
            print(f"Seat {self.seat_number} reserved.")
            return True
        else:
            print(f"Seat {self.seat_number} is already reserved.")
            return False

    # Method to unreserve the seat
    def unreserve_seat(self):
        if self.reserved:
            self.reserved = False
            print(f"Seat {self.seat_number} has been successfully unreserved.")
            return True
        else:
            print(f"Seat {self.seat_number} is not currently reserved.")
            return False

    # Method to check if the seat is reserved
    def is_reserved(self):
        return self.reserved


class SeatManager:
    def __init__(self, number_of_seats):
        self.seats = {i: Seat(i) for i in range(1, number_of_seats + 1)}  # Creating seat objects

    # Method to allocate a seat only if it is available
    def allocate_seat(self, seat_number):
        seat = self.seats.get(seat_number)
        if seat:
            if not seat.is_reserved():
                seat.reserve_seat()  # Allocate the seat if available
            else:
                print(f"Seat {seat_number} is already reserved. Cannot allocate.")
        else:
            print(f"Invalid seat number: {seat_number}")

    # Method to unreserve a seat by seat number
    def unreserve_seat(self, seat_number):
        seat = self.seats.get(seat_number)
        if seat:
            seat.unreserve_seat()
        else:
            print(f"Invalid seat number: {seat_number}")

    # Method to check seat availability
    def is_seat_available(self, seat_number):
        seat = self.seats.get(seat_number)
        return seat is not None and not seat.is_reserved()


# Example usage
if __name__ == "__main__":
    # Initialize the SeatManager with 20 seats
    seat_manager = SeatManager(20)

    # Example: Trying to allocate seats
    seat_manager.allocate_seat(5)  # Should reserve seat 5
    seat_manager.allocate_seat(5)  # Should show seat 5 is already reserved
    seat_manager.allocate_seat(10)  # Should reserve seat 10

    # Unreserve a seat
    seat_manager.unreserve_seat(5)  # Should unreserve seat 5
    seat_manager.allocate_seat(5)  # Should reserve seat 5 again after unreserving

    # Trying to reserve an invalid seat
    seat_manager.allocate_seat(21)  # Should print invalid seat number
